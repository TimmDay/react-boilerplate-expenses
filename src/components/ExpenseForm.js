import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates'; // requires moment


// this one is reusable.
// for use in AddExpense and EditExpense

// form fields, validation and logic

export default class ExpenseForm extends React.Component {


    // when form submitted, we will send it off to redux
    // to either edit existing expense or create new one
    //note optional, number can be zero if they want. just REQUIRE desc
    // define onCHange otherwise read only inputs

    // for amount input, enforce format
    // - only numbers, one decimal, two num after
    // - use string and convert to num last
    // conditional logic in oncChnage handler


    constructor(props) {
        super(props);
        // local component state
        this.state = {
            description: props.existingExpense ? props.existingExpense.description : '',
            note: props.existingExpense ? props.existingExpense.note : '',
            amount: props.existingExpense ? (props.existingExpense.amount / 100).toString() : '',
            createdAt: props.existingExpense ? moment(props.existingExpense.createdAt) : moment(), // current time
            calendarFocused: false,
            error: ''
        };
    }


    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description: description }));
    };
    onNoteChange = (e) => {
        // e.persist();
        this.setState(() => ({ note: e.target.value}));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount: amount }));
        }
    };


    onDateChange = (createdAt) => {
        if (createdAt) { //prevents user clearing the value
            // because if date is cleared, this func gets called with nothing
            // (we don't want to update this part of state to nothing)
            this.setState(() => ({ createdAt: createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(()=> ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault(); //otherwise browser does full page refresh

        if (!this.state.description || !this.state.amount) {
            // set error state equal 'please provide description and amount'
            this.setState(() => ({error: 'please provide description and amount'}))
        } else {
            this.setState(() => ({error: ''}));
            // the dispatch to store func, added as prop in AddExpense when form instantiated
            // * 100 to cents
            // javascript works in ms
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };


    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form
                    onSubmit={this.onSubmit}
                >
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />

                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />

                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false} //every day available, past and future
                    />

                    <textarea
                        placeholder="add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>

                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}
