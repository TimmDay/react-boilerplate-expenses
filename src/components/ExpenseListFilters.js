import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from "../actions/filters";
import { DateRangePicker } from 'react-dates';


class ExpenseListFilters extends React.Component {

    // class-based components need 'this' keyword before props
    constructor(props) {
        super(props);
    }

    state = {
        calendarFocused: null
    };

    onDatesChange = ({ startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };

    onFocusChange = (calendarFocused) => {
      this.setState(() => ({
          calendarFocused: calendarFocused
      }));
    };

    render() {
        return (
            <div>
                <input type="text"
                       value = {this.props.filters.text}
                       onChange={(e) =>{
                           //use dispatch to update redux store according to keystrokes
                           this.props.dispatch(setTextFilter(e.target.value));
                           console.log(e.target.value); //todo
                       }}
                />

                <select
                    value = {this.props.filters.sortBy}
                    onChange={(e) => {
                        e.target.value === 'date' && this.props.dispatch(sortByDate());
                        e.target.value === 'amount' && this.props.dispatch(sortByAmount());
                    }}
                >
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>

                <DateRangePicker
                    startDate={this.props.filters.startDate} // momentPropTypes.momentObj or null,
                    // startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.props.filters.endDate} // momentPropTypes.momentObj or null,
                    // endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false} //back in time
                />

            </div>
        )

    }
}



// puts the state from the store into the props for this component
const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};


export default connect(mapStateToProps)(ExpenseListFilters);