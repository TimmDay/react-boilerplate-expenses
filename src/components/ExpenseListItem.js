import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ description = '', amount='', createdAt='', id='-1' } ) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>

        <p>
            {/*${amount / 100}*/}
            {numeral(amount / 100).format('$0,0.00')}
            -
            {moment(createdAt).format('ddd, Do MMM YYYY')}
        </p>


    </div>
);

export default ExpenseListItem;

// a connect() by itself just gives access to the dispatch
// export unconnected component for testing. no connected version here so default is fine