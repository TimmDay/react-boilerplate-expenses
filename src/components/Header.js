import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1> The Title! </h1>
        <ul>
            <li><NavLink to="/" activeClassName="is-active" exact={true}>home</NavLink></li>
            <li><NavLink to="/create" activeClassName="is-active" exact={false}>add expense</NavLink></li>
            <li><NavLink to="/help" activeClassName="is-active" exact={false}>help</NavLink></li>
            <NavLink to="/dogs" activeClassName="is-active" exact={false}>broken link</NavLink>
        </ul>
    </header>
);

export default Header;