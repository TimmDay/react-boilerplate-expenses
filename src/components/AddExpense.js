import React from 'react';
import { Link } from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
// we don't dispatch in expense form, because we want that to be reusable
// data is different dep on add/edit expense, so dispatch at that level
import { connect } from 'react-redux'; //connect so we can dispatch to store
import { startAddExpense } from "../actions/expenses";

// export so that we can test the unconnected version of the component
export class AddExpense extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                <h1>add expense</h1>

                <ExpenseForm
                    onSubmit={this.onSubmit}
                />

                <Link to="/dashboard">back</Link>
            </div>
        )
    }
}

// export default AddExpense;

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpense);
// first parenthesis for mapping state as props, 2nd the component
// with connect wired up, we can access props.dispatch

// components rendered inside react router get access to special props -> history.push
// push takes single string path arg
// switches over like link click (no full page refresh)