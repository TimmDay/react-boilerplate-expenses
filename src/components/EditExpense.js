import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from "../actions/expenses";


export class EditExpense extends React.Component {

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/dashboard');
    };

    onRemove = () => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/dashboard');
    };

    render() {
        return (
            <div>
                {/*<p>editing expense {this.props.match.params.id} </p>*/}

                {/* populate fields with existing values */}
                <ExpenseForm
                    existingExpense={this.props.expense}
                    onSubmit={this.onSubmit}
                />

                <button
                    onClick = {this.onRemove}
                >Remove</button>
            </div>
        )
    }
}

// give the component the current expense object as props
const mapStateToProps = (state, props) =>
{
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id;
        })
    }
};
// MORE CONCISE PATTERN OF ABOVE
//     ({
//     expense: state.expenses.find((expense) => {
//         return expense.id === props.match.params.id;
//     })
// });

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);