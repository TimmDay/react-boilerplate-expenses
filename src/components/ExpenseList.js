import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from './../selectors/expenses';

// use map to iterate over list and render each one
// export unconnected verion (for testing only)
export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-larger-screens">Expense</div>
            <div className="show-for-larger-screens">Amount</div>
        </div>

        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div>
                        <span className="list-item list-item--empty-msg">no expenses</span>
                    </div>
                ) : (
                    props.expenses.map((ex) => (
                        <ExpenseListItem
                            key={ex.id}
                            {...ex}
                        />
                    ))
                )
            }
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);

//state.expenses is the complete array. to filter it we need our selector

// HOC - to access redux store
// connect gives us a function back... so second parenthesis to use pass in component
// in the connect call we also specify what part of the store we want to access, with a func
// return an object


// long version of code, for clarity of understanding / reference
// const ConnectedExpenseList = connect((state) => {
//
//     return {
//         expenses : state.expenses
//     };
//
// })(ExpenseList);
//
//
// export default ConnectedExpenseList;

//alternative: export default connect ((....
