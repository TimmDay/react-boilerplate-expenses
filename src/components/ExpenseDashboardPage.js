import React from 'react';
import { Link } from 'react-router-dom';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
        <p> the dashboard component </p>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;