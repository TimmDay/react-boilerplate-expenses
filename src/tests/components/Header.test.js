// react-test-renderer
import React from 'react';
import { shallow } from 'enzyme'; // released by airbnb. is a full-featured renderer for testing
// import ReactShallowRenderer from 'react-test-renderer/shallow';

import { Header } from '../../components/Header';


test('should render Header correctly', () => {
    // shallow render the header
    const wrapper = shallow(<Header startLogout={() => { }} />);
    expect(wrapper).toMatchSnapshot();
    // will update snap (with u in test suite) then compares future renders to this
    // expect(wrapper.find('h1').text()).toBe(" The Title! ");
    // expect(wrapper.find('h1').length).toBe(1); // we have only 1 h1 tag
});

test('should call startLogout on button click', () => {
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout={startLogout} />);
    wrapper.find('button').simulate('click');

    expect(startLogout).toHaveBeenCalled();
});

//todo update readme
// yarn add enzyme
// enzyme-adaptor-react-16
// raf (request animation frame)