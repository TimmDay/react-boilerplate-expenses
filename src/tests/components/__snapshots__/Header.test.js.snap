// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <h1>
     The Title! 
  </h1>
  <ul>
    <li>
      <NavLink
        activeClassName="is-active"
        ariaCurrent="true"
        exact={true}
        to="/"
      >
        log in
      </NavLink>
    </li>
    <li>
      <NavLink
        activeClassName="is-active"
        ariaCurrent="true"
        exact={false}
        to="/dashboard"
      >
        dash
      </NavLink>
    </li>
    <li>
      <NavLink
        activeClassName="is-active"
        ariaCurrent="true"
        exact={false}
        to="/create"
      >
        add expense
      </NavLink>
    </li>
    <li>
      <NavLink
        activeClassName="is-active"
        ariaCurrent="true"
        exact={false}
        to="/help"
      >
        help
      </NavLink>
    </li>
    <NavLink
      activeClassName="is-active"
      ariaCurrent="true"
      exact={false}
      to="/dogs"
    >
      broken link
    </NavLink>
    <button
      onClick={[Function]}
    >
      Logout
    </button>
  </ul>
</header>
`;
